test/manipulation.test.js

```diff
@@ -768,6 +768,80 @@ describe('manipulation', function() {
     });
   });
 
+  describe('upsertWithWhere', function() {
+    var ds = getSchema();
+    var Post;
+    before('prepare "Post" model', function(done) {
+      Post = ds.define('Post', {
+        empid: { type: Number, id: true },
+        name: { type: String },
+        city: { type: String },
+      });
+      ds.automigrate('Post', done);
+    });
+
+    it('has an alias "patchOrCreateWithWhere"', function() {
+      StubUser.upsertWithWhere.should.equal(StubUser.patchOrCreateWithWhere);
+    });
+
+    it('should not allow upsert when data object is empty (where,data,options,cb)', function(done) {
+      options = undefined;
+      Post.upsertWithWhere({ name: 'John Lennon' }, {}, options,
+        function(err) {
+          err.message.should.equal('data object cannot be empty!');
+          done();
+        });
+    });
+
+    it('should create a new record when no matching instances are found ' +
+      '(where, data, options, cb)', function(done) {
+      Post.upsertWithWhere({ where: { city: 'Florida' }}, { name: 'Nick Carter', empid: 1, city: 'Florida' },
+        function(err, created) {
+          if (err) return done(err);
+          created.empid.should.equal(1);
+          created.name.should.equal('Nick Carter');
+          created.city.should.equal('Florida');
```

Comment:
[My other comment](https://github.com/strongloop/loopback-datasource-juggler/pull/1001/files#r71381340) from previous review applies in most of the test cases including this test.